Custom fields validation fix

ptagl opened the pull request:
This PR contains a fix for the custom fields management. In particular, the validation function was using the wrong bit endianness to check custom fields against the configuration (bit length) specified during sidechain creation.

Commit:
Fixed an issue with the validation of custom fields

The validation is performed by checking that a field element doesn't use more bits than declared during sidechain creation. The function performing such validation was handling bits with wrong endianness.

Commit:
Fixed failing Python tests due to malformed custom fields

i-Alex commented on src/util.cpp:
```diff
@@ -934,21 +935,15 @@ int GetNumCores()
-    return c;
+    // __builtin_clz() returns the number of leading zeros for an unsigned int, so we need to perform a
+    // conversion to use it for an unsigned char.
+    return __builtin_clz(inputByte) % (sizeof(unsigned int) * 8 - CHAR_BIT);
```
Seems that `__builtin_clz` function is compiler and platform dependent and works with `unsigned int` type. Is it safe to use it? With trailing zeros before, we had safer solution, as for me.

ptagl replied:
Implemented a new function with commit [db1a66f2005b9348c04ee9b06c1aef8907622837](https://github.com/HorizenOfficial/zen/commit/db1a66f2005b9348c04ee9b06c1aef8907622837).

I moved `__builtin_clz()` to a unit test to check the validity of the new implemented function.

i-Alex commented on src/gtest/test_libzendoo.cpp:
```diff
@@ -1973,4 +1973,27 @@ TEST(CctpLibrary, TestInvalidProofVkWhenOversized)
+
+TEST(CctpLibrary, TestCustomFieldsValidation)
+{
+    for (uint8_t i = 1; i < CHAR_BIT; i++)
```
Here we have all possible cases for up-to 1 byte size custom field.
Let's add also some cases with bigger config size in a separate test. I mean, assuming that field element is created  from int\long\full field element. So actually the cases which failed before the fix.

ptagl replied:
Increased unit tests coverage with commit [7267da211e5e0fad639ca9cea7d20a0c6658daf8](https://github.com/HorizenOfficial/zen/commit/7267da211e5e0fad639ca9cea7d20a0c6658daf8).

Commit:
Increased unit test coverage for custom fields validation

Commit:
Implemented a "custom" function to compute trailing zero bits in a byte

This has been done to increase compatibility and avoid using a built-in function available only on specific architecture.

alsala commented on src/gtest/test_libzendoo.cpp:
```diff
@@ -19,6 +19,15 @@
 
 using namespace blockchain_test_utils;
 
+// Random uint used to generate a random custom field
```
It would be useful add a note explaining why this FE is created in this way (see also TestRandomCustomFieldGeneration below)

ptagl replied:
Done with commit [304694d2ecefdfe23ee5750d1c44e1b4562ca5ba](https://github.com/HorizenOfficial/zen/pull/429/commits/304694d2ecefdfe23ee5750d1c44e1b4562ca5ba).

I've also added a check to be sure that the first bit of the random custom field is set to "1".

alsala commented on src/gtest/test_libzendoo.cpp:
```diff
@@ -1973,4 +1982,140 @@ TEST(CctpLibrary, TestInvalidProofVkWhenOversized)
+        ASSERT_EQ(__builtin_clz(n) % (sizeof(unsigned int) * 8 - CHAR_BIT), getLeadingZeroBitsInByte(n));
+    }
+}
+
```
Could be useful also add a unit test for the getBytesFromBits() api in util.cpp

ptagl replied:
Done with commit [304694d2ecefdfe23ee5750d1c44e1b4562ca5ba](https://github.com/HorizenOfficial/zen/pull/429/commits/304694d2ecefdfe23ee5750d1c44e1b4562ca5ba).

alsala commented on src/gtest/test_libzendoo.cpp:
```diff
@@ -1973,4 +1982,140 @@ TEST(CctpLibrary, TestInvalidProofVkWhenOversized)
+            // and if doesn't exceed the modulus, thus the last two bits cannot be set (j < 1 << 6).
+            if (j < 1 << i && j < 1 << 6)
+            {
+                EXPECT_TRUE(fe.IsValid());
```
This is correct, even if according to the implementation of GetFieldElement(config) the returned fe is always the null fe if the bits of cfg are not consistent (null fe is invalid anyway)

Commit:
Added a unit test for the GetBytesFromBits() utility function

Also added a comment to explain the reason behind the implementation of the TestRandomCustomFieldGeneration unit test.

alsala commented on src/gtest/test_libzendoo.cpp:
```diff
@@ -2004,6 +2018,41 @@ TEST(CctpLibrary, TestGetLeadingZeros)
+    int reminder = 0;
+
+    // Check that the function works properly with the "0" input
+    ASSERT_EQ(0, getBytesFromBits(0, reminder));
```
This api has an int for the number of bits, it could be useful to add also a oneliner for testing negative values

Commit:
Removed UT failing on Mac OS due to different rand() implementation

Also added a check to TestGetBytesFromBits for negative values.
